src/main/scala/com/fulcrumgenomics/util/zip/Zipper.scala

```diff
@@ -0,0 +1,840 @@
+    * read data into blocks and store them in this queue, and consumers will consume blocks from this queue. */
+  private val toConsumeQueue: CloseableBlockingQueue[Blocks] = new CloseableBlockingQueue[Blocks]()
+  /** The queue of blocks that have output data, and are ready to be written to the output.  The consumers will put
+    * data that has been compressed or decompressed in this queue, and the writer will consume blocks from this queue. */
+  private val toWriteQueue: CloseableBlockingQueue[Blocks] = new CloseableBlockingQueue[Blocks]()
+
+  private val reader = new Reader(
+    in                = in,
+    decompress        = decompress,
+    inQueue           = unusedQueue,
+    outQueue          = toConsumeQueue,
+    numBlocksToBuffer = numBlocks
+  )
+
+  private val consumers: Seq[Consumer] = Seq.range(0, numThreads, 1).map { i =>
+    val consumer    = Consumer(
+      compress      = !decompress,
+      inQueue       = toConsumeQueue,
+      outQueue      = toWriteQueue,
+      id = i,
+      numThreads    = numThreads)
+    consumer
+  }
+
+  private val writer = Writer(
+    out       = out,
+    compress  = !decompress,
+    inQueue   = toWriteQueue,
+    outQueue  = unusedQueue
+  )
+
+  /** Some exception from a child thread. */
+  private var threadException: Option[Throwable] = None
+
+  /** The exception handler for child threads. Any exception will stop this thread and store the exception in
+    * `threadException`. */
+  private val threadExceptionHandler = new UncaughtExceptionHandler {
+    def uncaughtException(th: Thread, ex: Throwable): Unit = {
+      threadException = Some(ex)
+      stopIt()
+    }
+  }
+
+  /** Create all the blocks upfront that we will use throughout. */
+  Stream.range(0, numBlocks, 1).foreach { i => unusedQueue.put(new Blocks(blockId = i, compress = !decompress)) }
+
+  /** Returns true if `isStopped` has been set, or all data has been read in, consumed, and written out. */
+  private def isDone: Boolean = {
+    val ret = if (isStopped)                    true // we've been stopped
+    else {
+      if (!reader.isStopped)                    false // reader has not completed
+      else if (toConsumeQueue.nonEmpty)         false // more blocks to consume
+      else if (toWriteQueue.nonEmpty)           false // more blocks to write
+      else if (unusedQueue.size() != numBlocks) false // not all blocks written out (since they should be returned)
+      else                                      true  // all done!
+    }
+    ret
+  }
+
+  def run(): Unit = {
+    // Create the threads
+    val threads = (Seq(reader, writer) ++ consumers).map { runnable => new Thread(runnable) }
+    threads.foreach { thread => thread.setUncaughtExceptionHandler(threadExceptionHandler) }
+
+    // start your engines
+    threads.foreach { thread => thread.start() }
+
+    while (!isDone) {
+      // let others do stuff for a bit
+      Thread.sleep(Zipper.TimeoutInMilliseconds)
```

Comment:
I figured you'd just run the reader or writer on this thread.  Since you're not doing that I would suggest a longer sleep than 1ms - 10-50ms wouldn't affect response time for a user.